src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutorExtensions.cs

```diff
@@ -31,7 +32,25 @@ public static async Task<IDelayedException> TryExecuteAsync(this IFunctionExecut
                     logger = loggerFactory.CreateLogger(LogCategories.CreateFunctionCategory(functionInstance.FunctionDescriptor.LogName));
                 }
 
-                functionResult = await executor.TryExecuteAsync(functionInstance, cancellationToken);
+                if (functionInstance is FunctionInstance instance)
+                {
+                    // Build retry context
+                    retryContext = new RetryContext
+                    {
+                        RetryCount = attempt,
+                        Exception = retryContext == null ? null : retryContext.Exception, // Pass exception from previous attempt
+                        Instance = functionInstance,
+                        MaxRetryCount = (functionInstance.FunctionDescriptor.RetryStrategy == null) ? 0 : functionInstance.FunctionDescriptor.RetryStrategy.MaxRetryCount
+                    };
+
+                    instance.RetryContext = retryContext;
+                    functionResult = await executor.TryExecuteAsync(instance, cancellationToken);
```

Comment:
Looks like this code can be simplified by removing the else block below and executing these two lines outside of this if.